lib/datadog/statsd/connection_cfg.rb

```diff
@@ -0,0 +1,74 @@
+module Datadog
+  class Statsd
+    class ConnectionCfg
+      attr_reader :host
+      attr_reader :port
+      attr_reader :socket_path
+      attr_reader :transport_type
+
+      def initialize(host: nil, port: nil, socket_path: nil)
+        initialize_with_constructor_args(host: host, port: port, socket_path: socket_path) ||
+          initialize_with_env_vars ||
+          initialize_with_defaults
+      end
+
+      def make_connection(**params)
+        case @transport_type
+        when :udp
+          UDPConnection.new(@host, @port, **params)
+        when :uds
+          UDSConnection.new(@socket_path, **params)
+        end
+      end
+
+      private
+
+      DEFAULT_HOST = '127.0.0.1'
+      DEFAULT_PORT = 8125
+
+      def initialize_with_constructor_args(host: nil, port: nil, socket_path: nil)
+        try_initialize_with(host: host, port: port, socket_path: socket_path,
+          not_both_error_message: 
+            "Both UDP: (host/port %s:%s) and UDS (socket_path %s) constructor arguments were given.  Use only one or the other." %
+            [host, port, socket_path],
```

Comment:
Minor: It's possible, but not very common to use this kind of string templating in Ruby -- usually Ruby codebases use the inline style, e.g. `"Value of foo is #{foo}, value of bar is #{bar}"`

Reply:
Ah, I thought I remembered something like that!  It's been a looong time since I've Rubied, and my DDG search led me astray.  Thanks :D